ta/pkcs11/src/object.c

```diff
@@ -0,0 +1,408 @@
+/*
+ * Copyright (c) 2017-2020, Linaro Limited
+ */
+
+#include <inttypes.h>
+#include <string_ext.h>
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+
+#include "attributes.h"
+#include "handle.h"
+#include "object.h"
+#include "pkcs11_attributes.h"
+#include "pkcs11_helpers.h"
+#include "pkcs11_token.h"
+#include "sanitize_object.h"
+#include "serializer.h"
+
+struct pkcs11_object *pkcs11_handle2object(uint32_t handle,
+					   struct pkcs11_session *session)
+{
+	return handle_lookup(&session->object_handle_db, handle);
+}
+
+uint32_t pkcs11_object2handle(struct pkcs11_object *obj,
+			      struct pkcs11_session *session)
+{
+	return handle_lookup_handle(&session->object_handle_db, obj);
+}
+
+/* Currently handle pkcs11 sessions and tokens */
+
+static struct object_list *get_session_objects(void *session)
+{
+	/* Currently supporting only pkcs11 session */
+	struct pkcs11_session *ck_session = session;
+
+	return pkcs11_get_session_objects(ck_session);
+}
+
+static struct ck_token *get_session_token(void *session)
+{
+	/* Currently supporting only pkcs11 session */
+	struct pkcs11_session *ck_session = session;
+
+	return pkcs11_session2token(ck_session);
+}
+
+/* Release non-persistent resources of an object */
+static void cleanup_volatile_obj_ref(struct pkcs11_object *obj)
+{
+	if (!obj)
+		return;
+
+	if (obj->key_handle != TEE_HANDLE_NULL)
+		TEE_FreeTransientObject(obj->key_handle);
+
+	if (obj->attribs_hdl != TEE_HANDLE_NULL)
+		TEE_CloseObject(obj->attribs_hdl);
+
+	TEE_Free(obj->attributes);
+	TEE_Free(obj->uuid);
+	TEE_Free(obj);
+}
+
+/* Release resources of a persistent object including volatile resources */
+static void cleanup_persistent_object(struct pkcs11_object *obj,
+				      struct ck_token *token)
+{
+	TEE_Result res;
+
+	if (!obj)
+		return;
+
+	/* Open handle with write properties to destroy the object */
+	if (obj->attribs_hdl != TEE_HANDLE_NULL)
+		TEE_CloseObject(obj->attribs_hdl);
+
+	res = TEE_OpenPersistentObject(TEE_STORAGE_PRIVATE,
+				       obj->uuid, sizeof(TEE_UUID),
+				       TEE_DATA_FLAG_ACCESS_WRITE_META,
+				       &obj->attribs_hdl);
+	assert(!res);
+	if (res)
+		goto out;
+
+	TEE_CloseAndDeletePersistentObject1(obj->attribs_hdl);
+
+out:
+	obj->attribs_hdl = TEE_HANDLE_NULL;
+	destroy_object_uuid(token, obj);
+
+	LIST_REMOVE(obj, link);
+
+	cleanup_volatile_obj_ref(obj);
+}
+
+/*
+ * destroy_object - destroy an PKCS11 TA object
+ *
+ * @session - session requesting object destruction
+ * @object - reference to the PKCS11 TA object
+ * @session_only - true if only session object shall be destroyed
+ */
+void destroy_object(struct pkcs11_session *session, struct pkcs11_object *obj,
+		    bool session_only)
+{
+#ifdef DEBUG
+	trace_attributes("[destroy]", obj->attributes);
+	if (obj->uuid)
+		MSG_RAW("[destroy] obj uuid %pUl", (void *)obj->uuid);
+#endif
+
+	/* Remove from session list only if was published */
+	if (obj->link.le_next || obj->link.le_prev)
```

Comment:
I think it does because `obj->link.le_prev` is a pointer to the previous next pointer, in the case this is the only object it's a pointer to the `lh_first` pointer in the list head.

It is a bit ugly though.

Reply:
Please double-check and put a comment. Or maybe this can get a test is xtest.

Reply:
I've checked it with GDB. I'll add a comment.